clippy_lints/src/restricted_locales.rs

```diff
@@ -0,0 +1,408 @@
+use clippy_utils::diagnostics::span_lint;
+use rustc_data_structures::fx::FxHashSet;
+use rustc_hir::{Pat, PatKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_tool_lint, impl_lint_pass};
+use serde::Deserialize;
+use unicode_script::{Script, UnicodeScript};
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for usage of unicode scripts other than explicitly allowed
+    /// by the lint config.
+    ///
+    /// This lint doesn't take into account non-text scripts such as `Unknown` and `Linear_A`.
+    /// It also ignores the `Common` script type.
+    /// While configuring, be sure to use official script names from [the list of supported scripts][supported_scripts]
+    /// and replace spaces with undescores.
+    ///
+    /// For example, `Syloti Nagri` must be represented as `"Syloti_Nagri`".
+    ///
+    /// [supported_scripts]: http://www.unicode.org/standard/supported.html
+    ///
+    /// **Why is this bad?** It may be not desired to have many different locales for
+    /// identifiers in the codebase.
+    ///
+    /// Note that if you only want to allow plain English, you might want to use
+    /// built-in `non_ascii_idents` lint instead.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    /// ```rust
+    /// // Assuming that `clippy.toml` contains the following line:
+    /// // allowed-locales = ["Latin", "Cyrillic"]
+    /// let counter = 10; // OK, latin is allowed.
+    /// let счётчик = 10; // OK, cyrillic is allowed.
+    /// let zähler = 10; // OK, it's still latin.
+    /// let カウンタ = 10; // Will spawn the lint.
+    /// ```
+    pub RESTRICTED_LOCALES,
+    pedantic,
+    "usage of non-allowed locales"
+}
+
+#[derive(Clone, Debug)]
+pub struct RestrictedLocales {
+    whitelist: FxHashSet<AllowedScript>,
+}
+
+impl RestrictedLocales {
+    pub fn new(whitelist: FxHashSet<AllowedScript>) -> Self {
+        Self { whitelist }
+    }
+}
+
+impl_lint_pass!(RestrictedLocales => [RESTRICTED_LOCALES]);
+
+impl<'tcx> LateLintPass<'tcx> for RestrictedLocales {
+    fn check_pat(&mut self, cx: &LateContext<'tcx>, pat: &'tcx Pat<'_>) {
```

Comment:
Given that lint passes are always run, can you check if there's a measurable perf difference when linting a decently sized crate?

Reply:
It seems that I cannot just build the release version of clippy and run it. `clippy-driver` fails with the following error: `dyld: Library not loaded: @rpath/librustc_driver-22f067a38d2152ee.dylib`.

If I understood things correctly from looking through this repo issues, I need the whole rust toolchain for that, which is not desirable. Is there a simpler way to run clippy on a custom crate?

Reply:
You need to have the rustc-dev component installed. You're probably using a different toolchain when you build clippy vs when you run it.

Check out https://github.com/rust-lang/rust-clippy/blob/master/rust-toolchain#L3

Reply:
Dunno. The compiler is (and was) overridden to be the same as for clippy itself. Components are also seem to be installed.

```sh
$ rustup +nightly-2021-06-17 component add llvm-tools-preview 
info: component 'llvm-tools-preview' for target 'aarch64-apple-darwin' is up to date
$ rustup +nightly-2021-06-17 component add rust-src          
info: component 'rust-src' is up to date
$ rustup +nightly-2021-06-17 component add rustc-dev         
info: component 'rustc-dev' for target 'aarch64-apple-darwin' is up to date
$ rustup show
# ...
nightly-2021-06-17-aarch64-apple-darwin (directory override for '.../actix-web')
# ...
```

Yet it still fails:

```sh
$ ../rust-clippy/target/release/clippy-driver --deny clippy::needless_borrow
dyld: Library not loaded: @rpath/librustc_driver-22f067a38d2152ee.dylib
  Referenced from: /<...>/actix-web/../rust-clippy/target/release/clippy-driver
  Reason: image not found
[1]    32340 abort      ../rust-clippy/target/release/clippy-driver --deny clippy::needless_borrow
```

If that's important, I build clippy via simple `cargo build --release` (maybe I need to do it some other way?).